crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+/// Python code that we are evaluating. We intern these constraints in an arena
+/// ([`VisibilityConstraints::constraints`]). An atom consists of an index into this arena, and a
+/// copy number.
+#[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
+struct Atom(u32);
+
+impl Atom {
+    /// Deconstruct an atom into a constraint index and a copy number.
+    #[inline]
+    fn into_index_and_copy(self) -> (u32, u8) {
+        let copy = self.0 >> 24;
+        let index = self.0 & 0x00ff_ffff;
+        (index, copy as u8)
+    }
+
+    #[inline]
+    fn from_index_and_copy(index: u32, copy: u8) -> Self {
+        debug_assert!(index <= 0x00ff_ffff);
+        Self((u32::from(copy) << 24) | index)
+    }
 }
 
-/// A newtype-index for a visibility constraint in a particular scope.
-#[newtype_index]
-pub(crate) struct ScopedVisibilityConstraintId;
+// A custom Debug implementation that prints out the constraint index and copy number as distinct
+// fields.
+impl std::fmt::Debug for Atom {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (index, copy) = self.into_index_and_copy();
+        f.debug_tuple("Atom").field(&index).field(&copy).finish()
+    }
+}
 
 impl ScopedVisibilityConstraintId {
     /// A special ID that is used for an "always true" / "always visible" constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 0.
     pub(crate) const ALWAYS_TRUE: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(0);
+        ScopedVisibilityConstraintId(0xffff_ffff);
+
+    /// A special ID that is used for an ambiguous constraint.
+    pub(crate) const AMBIGUOUS: ScopedVisibilityConstraintId =
+        ScopedVisibilityConstraintId(0xffff_fffe);
 
     /// A special ID that is used for an "always false" / "never visible" constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 1.
     pub(crate) const ALWAYS_FALSE: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(1);
+        ScopedVisibilityConstraintId(0xffff_fffd);
 
-    /// A special ID that is used for an ambiguous constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 2.
-    pub(crate) const AMBIGUOUS: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(2);
+    fn is_terminal(self) -> bool {
+        self.0 >= SMALLEST_TERMINAL
+    }
 }
 
+// Rebind some constants locally so that we don't need as many qualifiers below.
+const ALWAYS_TRUE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_TRUE;
+const AMBIGUOUS: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::AMBIGUOUS;
+const ALWAYS_FALSE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_FALSE;
+const SMALLEST_TERMINAL: u32 = ALWAYS_FALSE.0;
```

Comment:
I'd expected this to also be defined on `ScopedVisiblityConstraitId` as it is used in `is_terminal` and that there's only an alias here.